Socrata export ETL

johnclary opened the pull request:
## Associated issues

- https://github.com/cityofaustin/atd-data-tech/issues/18149

This is the new Socrata export! The diff is not as bad as it looks—I had to renumber a bunch of migrations.

These follow-up tasks will be addressed in separate issues/PRs:

- airflow
- docker image CI
- prepare/coordinate updates to the production datasets to make them match the changes made here


## Testing

You can run this using a local db, but we're going to use the test instance cluster since it is already loaded up with 10 years of data.

### Setup

1. In our password store, locate the item called `Env file for the Vision Zero new data model Socrata export ETL` and save it in the root of this ETL directory as `.env`.

2. Build and run the Socrata export. This may take up to 20 minutes.

```shell
# from ./atd-etl/socrata_export
$ docker compose run socrata_export
$ ./socrata_export.py --crashes --people
```

3. Nice! Start your local VZV and inspect the summary page and map

```
# from ./atd-vzv
$ nvm use
$ npm run start
```

4. Test out the helper script, which logs the differences between the staging datasets and production datasets, which will need to be brought into sync with the staging datasets when we launch.

```shell
# from ./atd-etl/socrata_export
$ docker compose run socrata_export
$ ./_compare_column_metadata.py
```

johnclary commented on atd-etl/socrata_export/_compare_column_metadata.py:
This script identifies the differences between the staging and prod datasets and prints it in a nice way, like this:

**** crashes dataset changes ****

- change `crash_fatal_fl` from type `text` to `checkbox` in the prod dataset
- change `onsys_fl` from type `text` to `checkbox` in the prod dataset
- change `private_dr_fl` from type `text` to `checkbox` in the prod dataset
- change `road_constr_zone_fl` from type `text` to `checkbox` in the prod dataset
- create column `address_primary` type `text` in the prod dataset
- create column `address_secondary` type `text` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `law_enf_fatality_count` type `number` in the prod dataset
- create column `vz_fatality_count` type `number` in the prod dataset
- delete column `apd_confirmed_death_count` from the prod dataset
- delete column `apd_confirmed_fatality` from the prod dataset
- delete column `bicycle_fl` from the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `micromobility_fl` from the prod dataset
- delete column `motor_vehicle_fl` from the prod dataset
- delete column `motorcycle_fl` from the prod dataset
- delete column `other_fl` from the prod dataset
- delete column `pedestrian_fl` from the prod dataset
- delete column `rpt_latitude` from the prod dataset
- delete column `rpt_longitude` from the prod dataset
- delete column `street_name_2` from the prod dataset
- delete column `street_name` from the prod dataset
- delete column `street_nbr_2` from the prod dataset
- delete column `street_nbr` from the prod dataset

**** people dataset changes ****

- change `mode_id` from type `text` to `number` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `is_primary_person` type `checkbox` in the prod dataset
- create column `prsn_ethnicity_label` type `text` in the prod dataset
- create column `prsn_sex_id` type `number` in the prod dataset
- create column `prsn_sex_label` type `text` in the prod dataset
- create column `unit_id` type `number` in the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `prsn_gndr_id` from the prod dataset

johnclary commented on atd-toolbox/data_model/python/create_migrations.sh:
i finally created this helper to run all the python sql-gen scripts 👍

johnclary commented on atd-toolbox/data_model/python/40_make_lookup_seeds.py:
```diff
@@ -15,8 +16,8 @@ def escape_single_quotes(input_string):
-mode_category_lkp_values = """insert into lookups.mode_category_lkp (id, label)
-    select  id, atd_mode_category_mode_name from atd__mode_category_lkp;"""
+mode_category_lkp_values = """insert into lookups.mode_category_lkp (id, label, source)
+    select  id, atd_mode_category_mode_name, 'vz' from atd__mode_category_lkp where id != 10;"""
```
making it super clear that this lookup table is completely custom by adding a `source` column and setting it to `vz`. (we use the `source` column in other lookup tables which mix CRIS + VZ values—we have a documentation todo about this 👍 )

johnclary commented on atd-toolbox/data_model/python/column_metadata.json:
renamed `atd_mode_category ` to `vz_mode_category_id`

johnclary commented on atd-toolbox/data_model/python/column_metadata.json:
```diff
@@ -2262,19 +2262,19 @@
     "foreign_table_schema": "lookups",
     "is_cris_column": false,
-    "is_edits_column": true,
+    "is_edits_column": false,
```
it's a generated column, so we don't want it in the `_edits` schema

johnclary commented on atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:
summary of changes to this file:
- fix up some joining on `crashes.id` versus `crashes.crash_id`
- add the mode-based injury metric columns, which are rather tedious but we use them in the VZV
- use `LATERAL JOIN`s in the `crashes_list_view`, which turns out to be the silver bullet for keeping these queries as efficient as possible

johnclary commented on atd-vze/.env.test:
new env configuration for the test instance 🧑‍🔬

johnclary commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
+        records_processed = 0
+        while True:
+            logger.info(f"Fetching up to {RECORD_BATCH_SIZE} records")
+            variables = {"limit": RECORD_BATCH_SIZE, "minId": min_id}
```
i initially implemented this using the `offset` flag, but it was extremely slow. querying by `id` with a record limit is much faster. 

interestingly, i also had to push the `order by id` clause down into the views. this was dramatically faster than querying the view with an `order` directive. it's a limitation of the pg query planner that i don't understand.

johnclary commented on atd-etl/socrata_export/socrata_export.py:
as we're doing today, this script truncates and reloads each dataset in full every time it runs. that's ~150,000 crash records nightly 😵 

i think we could switch the crash level dataset over to incremental loading and it would work reliably. the person-level data is a bit more challenging because we only publish records for serious injuries and fatalities, so we would need a mechanism to remove records from person dataset. we could also switch over to publishing all person records. regardless of how we implement incremental loading, i think we would need some intermittent process that does a full truncate/replace to make sure we're 100% in sync.

TLDR i don't think we should fuss with incremental loading for the time being.

johnclary commented on atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:
summary of changes to this file:
- fix up some joining on `crashes.id` versus `crashes.crash_id`
- order views by `id`
- add the mode-based injury metric columns, which are rather tedious but we use them in the VZV
- use `LATERAL JOIN`s in the `crashes_list_view`, which turns out to be the silver bullet for keeping these queries fast

johnclary commented on atd-vze/.env.test:
new env configuration for the test instance 🧑‍🔬

johnclary commented on atd-etl/socrata_export/_compare_column_metadata.py:
the output of this script looks like this:

```**** crashes dataset changes ****

- change `crash_fatal_fl` from type `text` to `checkbox` in the prod dataset
- change `onsys_fl` from type `text` to `checkbox` in the prod dataset
- change `private_dr_fl` from type `text` to `checkbox` in the prod dataset
- change `road_constr_zone_fl` from type `text` to `checkbox` in the prod dataset
- create column `address_primary` type `text` in the prod dataset
- create column `address_secondary` type `text` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `id` type `number` in the prod dataset
- create column `law_enf_fatality_count` type `number` in the prod dataset
- delete column `apd_confirmed_death_count` from the prod dataset
- delete column `apd_confirmed_fatality` from the prod dataset
- delete column `atd_mode_category_metadata` from the prod dataset
- delete column `bicycle_fl` from the prod dataset
- delete column `contrib_factr_p1_id` from the prod dataset
- delete column `contrib_factr_p2_id` from the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `micromobility_fl` from the prod dataset
- delete column `motor_vehicle_fl` from the prod dataset
- delete column `motorcycle_fl` from the prod dataset
- delete column `other_fl` from the prod dataset
- delete column `pedestrian_fl` from the prod dataset
- delete column `rpt_latitude` from the prod dataset
- delete column `rpt_longitude` from the prod dataset
- delete column `street_name_2` from the prod dataset
- delete column `street_name` from the prod dataset
- delete column `street_nbr_2` from the prod dataset
- delete column `street_nbr` from the prod dataset

**** people dataset changes ****

- change `mode_id` from type `text` to `number` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `id` type `number` in the prod dataset
- create column `is_primary_person` type `checkbox` in the prod dataset
- create column `prsn_ethnicity_label` type `text` in the prod dataset
- create column `prsn_sex_id` type `number` in the prod dataset
- create column `prsn_sex_label` type `text` in the prod dataset
- create column `unit_id` type `number` in the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `prsn_gndr_id` from the prod dataset
```

roseeichelmann commented on atd-etl/socrata_export/_compare_column_metadata.py:
```diff
@@ -0,0 +1,59 @@
+        key = col["fieldName"]
+        matching_col = None
+        try:
+            matching_col = next(
```
this is a cool way to do this, lots of TILs in this python script for me :cowboy_hat_face:

roseeichelmann commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
-        },
-        "dataset_uid": ATD_ETL_CONFIG["SOCRATA_DATASET_CRASHES"],
+        "name": "crashes",
+        "typename": "socrata_export_crashes_view",
```
lovinggg the new socrata views

chiaberry replied:
+1

roseeichelmann commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
+                    raise Exception(
+                        "No records returned from Hasura. This should never happen"
+                    )
+                # we have some records, so it's safe to truncate the Socrata dataset
```
by truncate you mean dropping all the data in the table and replacing it?

roseeichelmann commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
-        client.upsert(config["dataset_uid"], records)
-        total_records += len(records)
+            logger.info(
+                f"Upserting {len(records)} records IDs: {records[0]['id']} - {records[-1]['id']} (total records: {records_processed})"
```
are we upserting tho or just completely replacing?

roseeichelmann replied:
jk i see we are upserting later i just got confused by how we replace at first?

chiaberry replied:
When min_id is 0 it replaces the whole dataset with an empty `[]` , and then does an upsert on every run until we run out of records, which is why the break happens on line 63. 

Could the logger say "(records already processed: {records_processed}" instead of "total records"? since its not the total number of records. 
`Upserting 1000 records IDS: 1-1000 (total records: 0)` is what I anticipate will log on the first run, right?

johnclary replied:
Thanks for suggesting the logging improvements. And I will add some comments to the code that make the logic for querying records and replacing the dataset more clear 🙏

roseeichelmann commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
 
-        # Format records
-        records = config["formatter"](data, config["formatter_config"])
+        min_id = 0
```
this code took me some time to understand, could you add some comments describing whats going on -- my understanding is we are looping thru querying the hasura database 1000 records at a time starting with a min record ID which gets incremented by +1 of the last record id at the end of the 1000 record batch replace. this was making me :woozy_face: trying to understand but maybe thats just me

johnclary replied:
will do! and yes, you got it :)

roseeichelmann commented on atd-etl/socrata_export/socrata_export.py:
```diff
@@ -1,121 +1,95 @@
+                    )
+                # we have some records, so it's safe to truncate the Socrata dataset
+                logger.info(f"Truncating dataset {dataset['dataset_id']}")
+                socrata_client.replace(dataset["dataset_id"], [])
```
im confused, why are we replacing the dataset when min_id = 0 and then afterwards we are upserting?

roseeichelmann commented on atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:
```diff
@@ -140,10 +261,46 @@ create or replace view crash_injury_metrics_view as
-            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 1
-            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 1
```
dont we want the `crash_injry_sev_id` to be representative of the most severe injury involved in the crash? so fatality -> serious injury -> nonincap injury etc? or am i misunderstanding this field

roseeichelmann commented on atd-toolbox/data_model/python/sql_templates/people_list_view.sql:
```diff
@@ -21,5 +21,5 @@ create or replace view people_list_view as (
     left join
         lookups.mode_category_lkp
-        on units.unit_desc_id = lookups.mode_category_lkp.id
+        on units.vz_mode_category_id = lookups.mode_category_lkp.id
```
great change :+1:

roseeichelmann commented on atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql:
```diff
@@ -0,0 +1,130 @@
+    limit 1
+) as unit_aggregates on true
+where
+    crashes.in_austin_full_purpose = true and crashes.private_dr_fl = false
```
should we be using this `private_dr_fl` like in the fatalities view as well?

roseeichelmann reviewed (approved):
this all looks soooo great john. tested with all your steps, it all seems to be working as expected. the helper script to compare the columns is a great touch. i just have a few questions and the socrata export script was a little hard for me to follow but i think i figured it out pretty well after taking time to comb into the code. exciting stuff!!! :rocket:

chiaberry commented on atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:
```diff
@@ -140,10 +261,46 @@ create or replace view crash_injury_metrics_view as
             when (sum(person_injury_metrics_view.fatal_injury) > 0) then 4
-            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 1
-            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 1
```
Im trying to follow whats happening, why did these swap?

chiaberry replied:
(I didnt see Rose's comment before I left mine, I need to refresh my tabs when im taking super long to review code)

johnclary replied:
ah yes—they swapped because injury severity ID `1`  = `sus_serious_injry`. so that was a bug. but as @roseeichelmann pointed out i introduced a new bug in my `case` statement 🙏

chiaberry commented on atd-etl/socrata_export/_compare_column_metadata.py:
```diff
@@ -0,0 +1,59 @@
+                todo = f"- change `{key}` from type `{b_col_type}` to `{a_col_type}` in the {b_list_name} dataset"
+                todos.append(todo)
+            else:
+                # we don't need to print these type mismatches
```
why doesnt it matter if the datatypes are different when the primary_list is b?

johnclary replied:
we only care about making the prod dataset look like the staging dataset, not the other way around. so there is no need to print anything. i am sorry this code is not easier to grok 😞

chiaberry commented on atd-etl/socrata_export/_compare_column_metadata.py:
```diff
@@ -0,0 +1,59 @@
+        cols_staging = socrata_client.get_metadata(dataset["staging"])["columns"]
+        cols_prod = socrata_client.get_metadata(dataset["prod"])["columns"]
+        todos = compare_col_arrays(cols_staging, cols_prod, "staging", "prod", primary_list="a")
+        todos += compare_col_arrays(cols_prod, cols_staging, "staging", "prod", primary_list="b")
```
I'm confused here

so the list_a and list_b lists are swapped, cols_prod is now list_a (when the line above here, cols_prod was list_b)

*but* a_list_name and b_list_name aren't swapped, so is it more like primary_list_name and secondary_list_name?

johnclary replied:
🍝

johnclary replied:
i'll touch up this code to try to make var names clearer

chiaberry commented:
I didn't run it, but read through the code and left a question or two. I got tripped up with the compare_column_metadata script a bit, but nothing major I don't think.

mddilley commented on atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql:
```diff
@@ -0,0 +1,130 @@
+) as unit_aggregates on true
+where
+    crashes.in_austin_full_purpose = true and crashes.private_dr_fl = false
+    and crashes.crash_timestamp < now() - interval '14 days' order by id asc;
```
love having this interval tucked into the database view. I know we touched on it long ago but it is cool to see this in action. Hooray for filtering in one place 🎉

mddilley reviewed (changes requested):
I tested this with the VZ test cluster [details here](https://austininnovation.slack.com/archives/CM9MK950S/p1720561270538219?thread_ts=1720559497.280309&cid=CM9MK950S), and I was able to fully replace both crash and people test datasets. I saw no obvious regressions in my local VZV either. It is so nice to see these refactored into database views. I know this will be more maintainable, and the sql really spells out the business logic well.

I found the tiniest typo that held up me doing a local run successfully, and I left some questions but I think this is 🚢 after the fix. 🙌

johnclary commented on atd-etl/socrata_export/utils/graphql.py:
```diff
@@ -0,0 +1,38 @@
+import requests
+
+ENDPOINT = os.getenv("HASURA_GRAPHQL_ENDPOINT")
+ADMIN_SECRET = os.getenv("HASURA_GRAPQL_ADMIN_SECRET")
```
nice catch—thank you! ✅

johnclary commented:
👋 thanks for the feedback. i reworked the logging and added docstrings in `socrata_export.py` and i refactored `_compare_column_metadata.py` to be less confusing 👍

mddilley reviewed (approved):
I just ran and replaced the crashes and people datasets again, and all worked perfectly!  I also diffed the output of the compare column metadata script, and it matches 100% before/after the refactor. The comments in that code make sense to me. 🙏 🚢 🚀
